Double or unnecessary Quoting of Columns in the filters

There appears to be an inconsistency in how column quoting is handled.

In Spark SQL, column names are already quoted when required (e.g., columns containing spaces or special characters) before they reach the clickhouse connector layer.

However, in the connector’s `SQLHelper.compileFilter` method, column names are always quoted unconditionally, regardless of whether they were already quoted by Spark.

This results in double quoting (e.g., space col → \``space col\`` )

**Suggested Fix** 
```
def quoted(token: String): String =
    if (token.startsWith("`") && token.endsWith("`")) token
    else s"`$token`"
```
**Schema of table**
```
1. │ CREATE TABLE default.space_col   ↴│
   │↳(                                ↴│
   │↳    `space col` Nullable(String),↴│
   │↳    `col` Int32 DEFAULT 1        ↴│
   │↳)                                ↴│
   │↳ENGINE = MergeTree               ↴│
   │↳ORDER BY tuple()                 ↴│
   │↳SETTINGS index_granularity = 8192 │
```
**Piece of code to Reproduce the issue:**
```
Dataset<Row> ds = spark.read()
            .format("clickhouse")
            ...
            .load();

ds.filter(functions.col("col").isNotNull()) .show(2, false); // Works Fine

ds.filter(functions.col("space col").isNotNull()).show(2, false);  // Fails when column contains space
```

**pom.xml**
```
<dependencies>
        <dependency>
            <groupId>com.clickhouse.spark</groupId>
            <artifactId>clickhouse-spark-runtime-3.5_2.12</artifactId>
            <version>0.10.0</version>
        </dependency>
        <dependency>
            <groupId>org.apache.spark</groupId>
            <artifactId>spark-core_2.12</artifactId>
            <version>3.5.0</version>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.spark</groupId>
            <artifactId>spark-sql_2.12</artifactId>
            <version>3.5.0</version>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-core</artifactId>
            <version>2.15.2</version>
        </dependency>
    </dependencies>
```

**Stacktrace**
```
Caused by: com.clickhouse.client.api.ServerException: Code: 62. DB::Exception: Syntax error: failed at position 274 ('``') (line 3, col 29): ``space col`` IS NOT NULL) AND (1=1))  LIMIT 3  . Expected one of: literal, NULL, NULL, number, Bool, TRUE, FALSE, string literal, SELECT query, possibly with UNION, list of union elements, SELECT query, subquery, possibly with UNION, SELECT subquery, SELECT query, WITH, FROM, SELECT, EXPLAIN, EXPLAIN, token, Comma, ClosingRoundBracket, CAST operator, NOT, INTERVAL, CASE, DATE, TIMESTAMP, tuple, collection of literals, array, asterisk, qualified asterisk, compound identifier, identifier, COLUMNS matcher, COLUMNS, qualified COLUMNS matcher, function name, substitution, MySQL-style global variable. (SYNTAX_ERROR) (version 25.1.1.4165 (official build))"] 
	at com.clickhouse.client.api.internal.HttpAPIClientHelper.readError(HttpAPIClientHelper.java:403)
	at com.clickhouse.client.api.internal.HttpAPIClientHelper.executeRequest(HttpAPIClientHelper.java:467)
	at com.clickhouse.client.api.Client.lambda$query$5(Client.java:1599)
	at com.clickhouse.client.api.Client.runAsyncOperation(Client.java:2012)
	at com.clickhouse.client.api.Client.query(Client.java:1644)
	at com.clickhouse.client.api.Client.query(Client.java:1542)
	at com.clickhouse.spark.client.NodeClient.$anonfun$queryAndCheck$2(NodeClient.scala:236)
	at scala.util.Try$.apply(Try.scala:210)
	at com.clickhouse.spark.client.NodeClient.queryAndCheck(NodeClient.scala:236)
	... 36 more

```
**Debugging Notes**
1. Verified via debugging `PushDownUtils.pushFilters` (Spark SQL) that:
1.1 Columns are already quoted only when necessary
1.2 Example: "col" → col  and "space col" → \`space col\`
2. By the time filters reach the connector, quoting has already been correctly applied
3. The issue arises when `SQLHelper.compileFilter` applies quoting again

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -122,6 +122,11 @@ Changes that have been merged but not yet released will be documented here.
 ### Removed
 - gzip and zstd write compression support has been removed (currently supported codecs are `none`, `lz4` (default))
 
+## [Unreleased]
+
+### Fixed
+- Preventing extra quotes in filter expr when col having special characters
+
 [Unreleased]: https://github.com/ClickHouse/spark-clickhouse-connector/compare/v0.9.0...HEAD
 [0.9.0]: https://github.com/ClickHouse/spark-clickhouse-connector/releases/tag/v0.9.0
 [0.8.1]: https://github.com/ClickHouse/spark-clickhouse-connector/releases/tag/v0.8.1

--- a/spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseReaderTestBase.scala
+++ b/spark-3.3/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseReaderTestBase.scala
@@ -14,7 +14,9 @@
 
 package org.apache.spark.sql.clickhouse.single
 
+import org.apache.spark.sql.functions.col
 import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.types._
 
 /**
  * Shared test cases for both JSON and Binary readers.
@@ -1491,6 +1493,41 @@ trait ClickHouseReaderTestBase extends SparkClickHouseSingleTest {
       if (!useSuiteLevelDatabase) {
         runClickHouseSQL(s"DROP DATABASE IF EXISTS $db")
       }
+    }
+  }
+
+  test("decode SpecialChar Cols: cols having space and applying filter") {
+    val schema: StructType = StructType(Seq(
+      StructField("id", IntegerType, nullable = false),
+      StructField("`space col`", StringType, nullable = true),
+      StructField("`1col`", StringType, nullable = true),
+      StructField("`hyphen-col`", StringType, nullable = true)
+    ))
+    withTable("test_db", "special_char_col", schema) {
+      (actualDb: String, actualTbl: String) =>
+        runClickHouseSQL(
+          s"""INSERT INTO $actualDb.special_char_col VALUES
+             |(1, NULL, NULL, NULL),
+             |(2, 'Bob', 'Bob2', 'Bob3')
+             |""".stripMargin
+        )
+
+        var df = readTableWithBothAPIs(
+          actualDb,
+          actualTbl,
+          columns = "id, `space col`, `1col`, `hyphen-col`",
+          orderBy = Some("id")
+        )
+        df = df.filter(col("space col").isNotNull)
+          .filter(col("1col").isNotNull)
+          .filter(col("hyphen-col").isNotNull)
+
+        val result = df.collect()
+        assert(result.length == 1)
+
+        assert(result(0).getString(1) == "Bob")
+        assert(result(0).getString(2) == "Bob2")
+        assert(result(0).getString(3) == "Bob3")
     }
   }
 

--- a/spark-3.3/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.3/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -16,6 +16,7 @@ package com.clickhouse.spark
 
 import java.sql.{Date, Timestamp}
 import java.time.{Instant, LocalDate, LocalDateTime, ZoneId}
+import java.util.regex.Pattern
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.connector.expressions.aggregate._
 import org.apache.spark.sql.connector.expressions.NamedReference
@@ -25,7 +26,13 @@ import com.clickhouse.spark.Utils._
 
 trait SQLHelper {
 
-  def quoted(token: String) = s"`$token`"
+  private val validIdentPattern = Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*")
+
+  private def alreadyQuoted(part: String): Boolean =
+    !validIdentPattern.matcher(part).matches()
+
+  def quoted(token: String): String =
+    if (alreadyQuoted(token)) token else s"`$token`"
 
   // null => null, ' => ''
   def escapeSql(value: String): String = StringUtils.replace(value, "'", "''")

--- a/spark-3.3/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.3/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -26,13 +26,10 @@ import com.clickhouse.spark.Utils._
 
 trait SQLHelper {
 
-  private val validIdentPattern = Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*")
-
-  private def alreadyQuoted(part: String): Boolean =
-    !validIdentPattern.matcher(part).matches()
-
-  def quoted(token: String): String =
-    if (alreadyQuoted(token)) token else s"`$token`"
+  def quoted(token: String): String = token match {
+    case t if t.startsWith("`") && t.endsWith("`") => t
+    case t => s"`$t`"
+  }
 
   // null => null, ' => ''
   def escapeSql(value: String): String = StringUtils.replace(value, "'", "''")

--- a/spark-3.3/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.3/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -16,7 +16,6 @@ package com.clickhouse.spark
 
 import java.sql.{Date, Timestamp}
 import java.time.{Instant, LocalDate, LocalDateTime, ZoneId}
-import java.util.regex.Pattern
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.connector.expressions.aggregate._
 import org.apache.spark.sql.connector.expressions.NamedReference

--- a/spark-3.4/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseReaderTestBase.scala
+++ b/spark-3.4/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseReaderTestBase.scala
@@ -14,7 +14,9 @@
 
 package org.apache.spark.sql.clickhouse.single
 
+import org.apache.spark.sql.functions.col
 import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.types._
 
 /**
  * Shared test cases for both JSON and Binary readers.
@@ -1491,6 +1493,41 @@ trait ClickHouseReaderTestBase extends SparkClickHouseSingleTest {
       if (!useSuiteLevelDatabase) {
         runClickHouseSQL(s"DROP DATABASE IF EXISTS $db")
       }
+    }
+  }
+
+  test("decode SpecialChar Cols: cols having space and applying filter") {
+    val schema: StructType = StructType(Seq(
+      StructField("id", IntegerType, nullable = false),
+      StructField("`space col`", StringType, nullable = true),
+      StructField("`1col`", StringType, nullable = true),
+      StructField("`hyphen-col`", StringType, nullable = true)
+    ))
+    withTable("test_db", "special_char_col", schema) {
+      (actualDb: String, actualTbl: String) =>
+        runClickHouseSQL(
+          s"""INSERT INTO $actualDb.special_char_col VALUES
+             |(1, NULL, NULL, NULL),
+             |(2, 'Bob', 'Bob2', 'Bob3')
+             |""".stripMargin
+        )
+
+        var df = readTableWithBothAPIs(
+          actualDb,
+          actualTbl,
+          columns = "id, `space col`, `1col`, `hyphen-col`",
+          orderBy = Some("id")
+        )
+        df = df.filter(col("space col").isNotNull)
+          .filter(col("1col").isNotNull)
+          .filter(col("hyphen-col").isNotNull)
+
+        val result = df.collect()
+        assert(result.length == 1)
+
+        assert(result(0).getString(1) == "Bob")
+        assert(result(0).getString(2) == "Bob2")
+        assert(result(0).getString(3) == "Bob3")
     }
   }
 

--- a/spark-3.4/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.4/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -16,6 +16,7 @@ package com.clickhouse.spark
 
 import java.sql.{Date, Timestamp}
 import java.time.{Instant, LocalDate, LocalDateTime, ZoneId}
+import java.util.regex.Pattern
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.connector.expressions.aggregate._
 import org.apache.spark.sql.connector.expressions.NamedReference
@@ -25,7 +26,13 @@ import com.clickhouse.spark.Utils._
 
 trait SQLHelper {
 
-  def quoted(token: String) = s"`$token`"
+  private val validIdentPattern = Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*")
+
+  private def alreadyQuoted(part: String): Boolean =
+    !validIdentPattern.matcher(part).matches()
+
+  def quoted(token: String): String =
+    if (alreadyQuoted(token)) token else s"`$token`"
 
   // null => null, ' => ''
   def escapeSql(value: String): String = StringUtils.replace(value, "'", "''")

--- a/spark-3.4/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.4/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -26,13 +26,10 @@ import com.clickhouse.spark.Utils._
 
 trait SQLHelper {
 
-  private val validIdentPattern = Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*")
-
-  private def alreadyQuoted(part: String): Boolean =
-    !validIdentPattern.matcher(part).matches()
-
-  def quoted(token: String): String =
-    if (alreadyQuoted(token)) token else s"`$token`"
+  def quoted(token: String): String = token match {
+    case t if t.startsWith("`") && t.endsWith("`") => t
+    case t => s"`$t`"
+  }
 
   // null => null, ' => ''
   def escapeSql(value: String): String = StringUtils.replace(value, "'", "''")

--- a/spark-3.4/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.4/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -16,7 +16,6 @@ package com.clickhouse.spark
 
 import java.sql.{Date, Timestamp}
 import java.time.{Instant, LocalDate, LocalDateTime, ZoneId}
-import java.util.regex.Pattern
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.connector.expressions.aggregate._
 import org.apache.spark.sql.connector.expressions.NamedReference

--- a/spark-3.5/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseReaderTestBase.scala
+++ b/spark-3.5/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseReaderTestBase.scala
@@ -14,7 +14,9 @@
 
 package org.apache.spark.sql.clickhouse.single
 
+import org.apache.spark.sql.functions.col
 import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.types._
 
 /**
  * Shared test cases for both JSON and Binary readers.
@@ -1491,6 +1493,41 @@ trait ClickHouseReaderTestBase extends SparkClickHouseSingleTest {
       if (!useSuiteLevelDatabase) {
         runClickHouseSQL(s"DROP DATABASE IF EXISTS $db")
       }
+    }
+  }
+
+  test("decode SpecialChar Cols: cols having space and applying filter") {
+    val schema: StructType = StructType(Seq(
+      StructField("id", IntegerType, nullable = false),
+      StructField("`space col`", StringType, nullable = true),
+      StructField("`1col`", StringType, nullable = true),
+      StructField("`hyphen-col`", StringType, nullable = true)
+    ))
+    withTable("test_db", "special_char_col", schema) {
+      (actualDb: String, actualTbl: String) =>
+        runClickHouseSQL(
+          s"""INSERT INTO $actualDb.special_char_col VALUES
+             |(1, NULL, NULL, NULL),
+             |(2, 'Bob', 'Bob2', 'Bob3')
+             |""".stripMargin
+        )
+
+        var df = readTableWithBothAPIs(
+          actualDb,
+          actualTbl,
+          columns = "id, `space col`, `1col`, `hyphen-col`",
+          orderBy = Some("id")
+        )
+        df = df.filter(col("space col").isNotNull)
+          .filter(col("1col").isNotNull)
+          .filter(col("hyphen-col").isNotNull)
+
+        val result = df.collect()
+        assert(result.length == 1)
+
+        assert(result(0).getString(1) == "Bob")
+        assert(result(0).getString(2) == "Bob2")
+        assert(result(0).getString(3) == "Bob3")
     }
   }
 

--- a/spark-3.5/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.5/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -16,6 +16,7 @@ package com.clickhouse.spark
 
 import java.sql.{Date, Timestamp}
 import java.time.{Instant, LocalDate, LocalDateTime, ZoneId}
+import java.util.regex.Pattern
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.connector.expressions.aggregate._
 import org.apache.spark.sql.connector.expressions.NamedReference
@@ -25,7 +26,13 @@ import Utils._
 
 trait SQLHelper {
 
-  def quoted(token: String) = s"`$token`"
+  private val validIdentPattern = Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*")
+
+  private def alreadyQuoted(part: String): Boolean =
+    !validIdentPattern.matcher(part).matches()
+
+  def quoted(token: String): String =
+    if (alreadyQuoted(token)) token else s"`$token`"
 
   // null => null, ' => ''
   def escapeSql(value: String): String = StringUtils.replace(value, "'", "''")

--- a/spark-3.5/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.5/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -26,13 +26,10 @@ import Utils._
 
 trait SQLHelper {
 
-  private val validIdentPattern = Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*")
-
-  private def alreadyQuoted(part: String): Boolean =
-    !validIdentPattern.matcher(part).matches()
-
-  def quoted(token: String): String =
-    if (alreadyQuoted(token)) token else s"`$token`"
+  def quoted(token: String): String = token match {
+    case t if t.startsWith("`") && t.endsWith("`") => t
+    case t => s"`$t`"
+  }
 
   // null => null, ' => ''
   def escapeSql(value: String): String = StringUtils.replace(value, "'", "''")

--- a/spark-3.5/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-3.5/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -16,7 +16,6 @@ package com.clickhouse.spark
 
 import java.sql.{Date, Timestamp}
 import java.time.{Instant, LocalDate, LocalDateTime, ZoneId}
-import java.util.regex.Pattern
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.connector.expressions.aggregate._
 import org.apache.spark.sql.connector.expressions.NamedReference

--- a/spark-4.0/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseReaderTestBase.scala
+++ b/spark-4.0/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseReaderTestBase.scala
@@ -14,6 +14,7 @@
 
 package org.apache.spark.sql.clickhouse.single
 
+import org.apache.spark.sql.functions.col
 import org.apache.spark.sql.{DataFrame, Row}
 import org.apache.spark.sql.types._
 
@@ -2013,6 +2014,41 @@ trait ClickHouseReaderTestBase extends SparkClickHouseSingleTest {
 
       val json3 = variantToJson(result(2).get(1).asInstanceOf[org.apache.spark.unsafe.types.VariantVal])
       assert(json3.contains("2.718"))
+    }
+  }
+
+  test("decode SpecialChar Cols: cols having space and applying filter") {
+    val schema: StructType = StructType(Seq(
+      StructField("id", IntegerType, nullable = false),
+      StructField("`space col`", StringType, nullable = true),
+      StructField("`1col`", StringType, nullable = true),
+      StructField("`hyphen-col`", StringType, nullable = true)
+    ))
+    withTable("test_db", "special_char_col", schema) {
+      (actualDb: String, actualTbl: String) =>
+        runClickHouseSQL(
+          s"""INSERT INTO $actualDb.special_char_col VALUES
+             |(1, NULL, NULL, NULL),
+             |(2, 'Bob', 'Bob2', 'Bob3')
+             |""".stripMargin
+        )
+
+        var df = readTableWithBothAPIs(
+          actualDb,
+          actualTbl,
+          columns = "id, `space col`, `1col`, `hyphen-col`",
+          orderBy = Some("id")
+        )
+        df = df.filter(col("space col").isNotNull)
+          .filter(col("1col").isNotNull)
+          .filter(col("hyphen-col").isNotNull)
+
+        val result = df.collect()
+        assert(result.length == 1)
+
+        assert(result(0).getString(1) == "Bob")
+        assert(result(0).getString(2) == "Bob2")
+        assert(result(0).getString(3) == "Bob3")
     }
   }
 

--- a/spark-4.0/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-4.0/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -16,6 +16,7 @@ package com.clickhouse.spark
 
 import java.sql.{Date, Timestamp}
 import java.time.{Instant, LocalDate, LocalDateTime, ZoneId}
+import java.util.regex.Pattern
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.connector.expressions.aggregate._
 import org.apache.spark.sql.connector.expressions.NamedReference
@@ -25,7 +26,13 @@ import Utils._
 
 trait SQLHelper {
 
-  def quoted(token: String) = s"`$token`"
+  private val validIdentPattern = Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*")
+
+  private def alreadyQuoted(part: String): Boolean =
+    !validIdentPattern.matcher(part).matches()
+
+  def quoted(token: String): String =
+    if (alreadyQuoted(token)) token else s"`$token`"
 
   // null => null, ' => ''
   def escapeSql(value: String): String = StringUtils.replace(value, "'", "''")

--- a/spark-4.0/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-4.0/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -26,13 +26,10 @@ import Utils._
 
 trait SQLHelper {
 
-  private val validIdentPattern = Pattern.compile("^[a-zA-Z_][a-zA-Z0-9_]*")
-
-  private def alreadyQuoted(part: String): Boolean =
-    !validIdentPattern.matcher(part).matches()
-
-  def quoted(token: String): String =
-    if (alreadyQuoted(token)) token else s"`$token`"
+  def quoted(token: String): String = token match {
+    case t if t.startsWith("`") && t.endsWith("`") => t
+    case t => s"`$t`"
+  }
 
   // null => null, ' => ''
   def escapeSql(value: String): String = StringUtils.replace(value, "'", "''")

--- a/spark-4.0/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
+++ b/spark-4.0/clickhouse-spark/src/main/scala/com/clickhouse/spark/SQLHelper.scala
@@ -16,7 +16,6 @@ package com.clickhouse.spark
 
 import java.sql.{Date, Timestamp}
 import java.time.{Instant, LocalDate, LocalDateTime, ZoneId}
-import java.util.regex.Pattern
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.sql.connector.expressions.aggregate._
 import org.apache.spark.sql.connector.expressions.NamedReference

